File: permissionsex-core/src/main/java/ninja/leaping/permissionsex/command/InfoCommand.java
/**
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.leaping.permissionsex.command;

import ninja.leaping.permissionsex.PermissionsEx;
import ninja.leaping.permissionsex.data.DataSegment;
import ninja.leaping.permissionsex.data.ImmutableSubjectData;
import ninja.leaping.permissionsex.data.SubjectRef;
import ninja.leaping.permissionsex.subject.CalculatedSubject;
import ninja.leaping.permissionsex.util.Translatable;
import ninja.leaping.permissionsex.util.Tristate;
import ninja.leaping.permissionsex.util.command.CommandContext;
import ninja.leaping.permissionsex.util.command.CommandException;
import ninja.leaping.permissionsex.util.command.CommandSpec;
import ninja.leaping.permissionsex.util.command.Commander;

import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Predicate;

import static ninja.leaping.permissionsex.util.Translations.t;

public class InfoCommand {
    public static CommandSpec getInfoCommand(PermissionsEx pex) {
        return CommandSpec.builder()
                .setAliases("info", "i", "who")
                .setDescription(t("Provide information about a subject"))
                .setExecutor(new SubjectInfoPrintingExecutor(pex))
                .build();
    }

    // TODO: Pagination builder

    private static class SubjectInfoPrintingExecutor extends PermissionsExExecutor {
        private static final String INDENT = "  ";
        private static final String DOUBLE_INDENT = INDENT + INDENT;

        private SubjectInfoPrintingExecutor(PermissionsEx pex) {
            super(pex);
        }

        @Override
        public <TextType> void execute(Commander<TextType> src, CommandContext args) throws CommandException {
            CalculatedSubject subject = subjectOrSelf(src, args);
            checkSubjectPermission(src, subject.getIdentifier(), "permissionsex.info");
            final ImmutableSubjectData transientData = subject.transientData().get();
            final ImmutableSubjectData data = subject.data().get();

            src.msg(src.fmt().header(src.fmt().tr(t("Information for %s", src.fmt().subject(subject)))));
            printSegments(src, t("Permissions:"), t("Transient permissions"), data, transientData,
                    seg -> !seg.getPermissions().isEmpty() || seg.getPermissionDefault() != Tristate.UNDEFINED,
                    (msg, seg) -> {
                        msg.accept(src.fmt().hl(src.fmt().tr(t("Default permission: %s", seg.getPermissionDefault()))));
                        for (Map.Entry<String, Tristate> ent : seg.getPermissions().entrySet()) {
                            msg.accept(src.fmt().permission(ent.getKey(), ent.getValue()));
                        }
                    });

            printSegments(src, t("Options:"), t("Transient options:"), data, transientData,
                     seg -> !seg.getOptions().isEmpty(),
                    (msg, seg) -> {
                        for (Map.Entry<String, String> option : seg.getOptions().entrySet()) {
                            msg.accept(src.fmt().option(option.getKey(), option.getValue()));
                        }
                    });

            printSegments(src, t("Parents:"), t("Transient parents:"), data, transientData,
                    seg -> !seg.getParents().isEmpty(),
                    (msg, seg) -> {
                        for (SubjectRef parent : seg.getParents()) {
                            msg.accept(src.fmt().subject(parent));
                        }
                    });
        }

        private <TextType> void printSegments(Commander<TextType> src, Translatable heading, Translatable transientHeading, ImmutableSubjectData data, ImmutableSubjectData transientData, Predicate<DataSegment> segmentNonEmpty, BiConsumer<Consumer<TextType>, DataSegment> print) {
            printSegmentsSingle(src, heading, data, segmentNonEmpty, print);
            printSegmentsSingle(src, transientHeading, transientData, segmentNonEmpty, print);
        }

        private <TextType> void printSegmentsSingle(Commander<TextType> src, Translatable heading, ImmutableSubjectData data, Predicate<DataSegment> segmentNonEmpty, BiConsumer<Consumer<TextType>, DataSegment> print) {
            boolean printed = false;
            for (DataSegment segment : data.getAllSegments()) {
                if (segmentNonEmpty.test(segment)) {
                    if (!printed) {
                        src.msg(src.fmt().hl(src.fmt().tr(heading)));
                        printed = true;
                    }
                    src.msg(src.fmt().combined(INDENT, formatSegmentKey(src, segment.getKey()), ":"));
                    print.accept(msg -> src.msg(src.fmt().combined(DOUBLE_INDENT)), segment);
                }
            }
        }
    }
}

File: core/src/main/kotlin/ca/stellardrift/permissionsex/commands/InfoCommand.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.commands

import ca.stellardrift.permissionsex.PermissionsEx
import ca.stellardrift.permissionsex.commands.Messages.INFO_ACTIVE_CONTEXTS
import ca.stellardrift.permissionsex.commands.Messages.INFO_ACTIVE_USED_CONTEXTS
import ca.stellardrift.permissionsex.commands.Messages.INFO_ASSOCIATED_OBJECT
import ca.stellardrift.permissionsex.commands.Messages.INFO_DESCRIPTION
import ca.stellardrift.permissionsex.commands.Messages.INFO_HEADER
import ca.stellardrift.permissionsex.commands.Messages.INFO_HEADER_OPTIONS
import ca.stellardrift.permissionsex.commands.Messages.INFO_HEADER_OPTIONS_TRANSIENT
import ca.stellardrift.permissionsex.commands.Messages.INFO_HEADER_PARENTS_TRANSIENT
import ca.stellardrift.permissionsex.commands.Messages.INFO_HEADER_PERMISSIONS
import ca.stellardrift.permissionsex.commands.Messages.INFO_HEADER_PERMISSIONS_TRANSIENT
import ca.stellardrift.permissionsex.commands.Messages.INFO_PERMISSIONS_DEFAULT
import ca.stellardrift.permissionsex.commands.commander.Commander
import ca.stellardrift.permissionsex.commands.commander.MessageFormatter
import ca.stellardrift.permissionsex.commands.parse.CommandContext
import ca.stellardrift.permissionsex.commands.parse.CommandException
import ca.stellardrift.permissionsex.commands.parse.command
import ca.stellardrift.permissionsex.data.ImmutableSubjectData
import ca.stellardrift.permissionsex.util.TranslatableProvider
import ca.stellardrift.permissionsex.util.join
import ca.stellardrift.permissionsex.util.plus
import ca.stellardrift.permissionsex.util.toComponent
import ca.stellardrift.permissionsex.util.unaryPlus
import net.kyori.text.BuildableComponent
import net.kyori.text.Component
import net.kyori.text.ComponentBuilder

internal fun getInfoCommand(pex: PermissionsEx<*>) =
    command("info", "i", "who") {
        description = INFO_DESCRIPTION()
        executor(SubjectInfoPrintingExecutor(pex))
    }

fun <C : BuildableComponent<C, B>, B : ComponentBuilder<C, B>> MessageFormatter.hlKeyVal(
    key: B,
    value: Component
): Component {
    return key.hl().build() + value
}

// TODO: Pagination builder
internal class SubjectInfoPrintingExecutor constructor(pex: PermissionsEx<*>) : PermissionsExExecutor(pex) {
    @Throws(CommandException::class)
    override fun execute(src: Commander, args: CommandContext) {
        val subject = subjectOrSelf(src, args)
        src.checkSubjectPermission(subject.identifier, "permissionsex.info")
        val transientData = subject.transientData().get()
        val data = subject.data().get()
        src.msg { send ->
            fun head(msg: TranslatableProvider) {
                send(msg.get().hl().build())
            }
            send(INFO_HEADER[subject(subject)].header().build())
            if (pex.hasDebugMode()) {
                val associatedObject = subject.associatedObject
                if (associatedObject != null) {
                    send(INFO_ASSOCIATED_OBJECT.get().hl().append(associatedObject.toString()).build())
                }
            }
            send(hlKeyVal(INFO_ACTIVE_CONTEXTS.get(), +subject.activeContexts.toString()))
            send(hlKeyVal(INFO_ACTIVE_USED_CONTEXTS.get(), +subject.usedContextValues.join().toString()))

            if (data.allPermissions.isNotEmpty() || data.allDefaultValues.isNotEmpty()) {
                head(INFO_HEADER_PERMISSIONS)
                printPermissions(src, data)
            }
            if (transientData.allPermissions.isNotEmpty() || transientData.allDefaultValues.isNotEmpty()) {
                head(INFO_HEADER_PERMISSIONS_TRANSIENT)
                printPermissions(src, transientData)
            }
            if (data.allOptions.isNotEmpty()) {
                head(INFO_HEADER_OPTIONS)
                printOptions(src, data)
            }
            if (transientData.allOptions.isNotEmpty()) {
                head(INFO_HEADER_OPTIONS_TRANSIENT)
                printOptions(src, transientData)
            }
            if (data.allParents.isNotEmpty()) {
                head(Messages.INFO_HEADER_PARENTS)
                printParents(src, data)
            }
            if (transientData.allParents.isNotEmpty()) {
                head(INFO_HEADER_PARENTS_TRANSIENT)
                printParents(src, transientData)
            }
        }
    }

    private fun MessageFormatter.printPermissions(
        src: Commander,
        data: ImmutableSubjectData
    ) {
        val targetContexts = data.allPermissions.keys + data.allDefaultValues.keys
        for (entry in targetContexts) {
            src.msg(listOf(INDENT, entry.toComponent(), COLON).join(separator = null))
            src.msg(DOUBLE_INDENT + INFO_PERMISSIONS_DEFAULT[data.getDefaultValue(entry).toComponent()].hl().build())
            data.getPermissions(entry).forEach { (k, v) ->
                src.msg(DOUBLE_INDENT + permission(k, v))
            }
        }
    }

    private fun MessageFormatter.printOptions(
        src: Commander,
        data: ImmutableSubjectData
    ) {

        for ((key, value) in data.allOptions) {
            src.msg(listOf(INDENT, key.toComponent(), COLON).join(separator = null))
            value.forEach { (k, v) ->
                src.msg(DOUBLE_INDENT + option(k, v))
            }
        }
    }

    private fun MessageFormatter.printParents(
        src: Commander,
        data: ImmutableSubjectData
    ) {
        for ((key, value) in data.allParents) {
            src.msg(listOf(INDENT, key.toComponent(), COLON).join(separator = null))
            for (parent in value) {
                src.msg(DOUBLE_INDENT + subject(parent))
            }
        }
    }

    companion object {
        private val NONE = +""
        private val COLON = +":"
        private val INDENT = +"  "
        private val DOUBLE_INDENT = +"    "
    }
}
